src/sasl/cyrsasl_scram.erl

```diff
@@ -58,164 +58,194 @@ mech_new(_Host, Creds) ->
-                                                   SaltedPassword =
-                                                   mongoose_scram:salted_password(Ret,
-                                                                                  TempSalt,
-                                                                                  mongoose_scram:iterations()),
-                                                   {mongoose_scram:stored_key(mongoose_scram:client_key(SaltedPassword)),
-                                                    mongoose_scram:server_key(SaltedPassword),
-                                                    TempSalt,
-                                                    mongoose_scram:iterations()}
-                                            end,
-                                            {NStart, _} = binary:match(ClientIn, <<"n=">>),
-                                            ClientFirstMessageBare = binary:part(ClientIn,
-                                                                                 {NStart,
-                                                                                  byte_size(ClientIn)-NStart}),
-                                            ServerNonce =
-                                            jlib:encode_base64(crypto:strong_rand_bytes(?NONCE_LENGTH)),
-                                            ServerFirstMessage =
-                                            iolist_to_binary(
-                                              [<<"r=">>,
-                                               ClientNonce,
-                                               ServerNonce,
-                                               <<",s=">>,
-                                               jlib:encode_base64(Salt),
-                                               <<",i=">>,
-                                               integer_to_list(IterationCount)]),
-                                            {continue, ServerFirstMessage,
-                                             State#state{step = 4, stored_key = StoredKey,
-                                                         server_key = ServerKey,
-                                                         auth_message =
-                                                         <<ClientFirstMessageBare/binary,
-                                                           ",", ServerFirstMessage/binary>>,
-                                                         client_nonce = ClientNonce,
-                                                         server_nonce = ServerNonce,
-                                                         username = UserName,
-                                                         auth_module = AuthModule}}
-                                    end;
-                                _Else -> {error, <<"not-supported">>}
-                            end
-                    end
+    ClientInList = re:split(ClientIn, <<",">>, [{return, binary}]),
```

Comment:
```suggestion
    ClientInList = binary:split(ClientIn, <<",">>, [global]),
```

So, I think we don't need a regex for something so simple, a binary split will do with much much better performance:
```erlang
10> Cs = [ crypto:strong_rand_bytes(100) || _ <- lists:seq(1,1000) ].
13> lists:all(fun(X) -> X == true end, [ re:split(C, <<",">>, [{return, binary}]) == binary:split(C, <<",">>, [global]) || C <- Cs ]).
true
14> lists:all(fun(X) -> X == true end, [ re:split(C, <<",">>, [{return, binary}]) == binary:split(C, <<",">>, [global]) || C <- Cs ]).
true
15> timer:tc(fun() ->   [re:split(C, <<",">>, [{return, binary}]) || C <- Cs], ok   end).
{8322,ok}
16> timer:tc(fun() ->   [binary:split(C, <<",">>, [global]) || C <- Cs], ok   end).
{3688,ok}
```
Or even in the case of a single element:
```erlang
21> C = crypto:strong_rand_bytes(100).
22> timer:tc(fun() ->   binary:split(C, <<",">>, [global]), ok   end).
{19,ok}
23> timer:tc(fun() ->   re:split(C, <<",">>, [{return, binary}]), ok   end).
{25,ok}
```

I believe this to work, as `mech_step` is called only from `cyrsasl:server_start/3`, which is called from `ejabberd_c2s` by passing something that we can be sure is a binary and nothing else, as it simply is the output of `base64:mime_decode/1`.

Would be cool to make sure nevertheless that I'm passing the correct options to the `binary:split/3` function, but that could be as easy as just throwing it to travis 😛

Reply:
I changed to the binary:split with the global option and it works locally.